miden-lib/asm/miden/note.masm

```diff
@@ -129,3 +129,110 @@ export.get_sender
     syscall.get_note_sender
     # => [sender]
 end
+
+#! Computes hash of note inputs starting at the specified memory address.
+#!
+#! Inputs:  [inputs_ptr, num_inputs]
+#! Outputs: [HASH]
+#!
+#! Panics if num_inputs is greater than 128.
+export.compute_inputs_hash
+    swap
+    # => [num_inputs, inputs_ptr]
+
+    # check that number of inputs is less than 128
+    dup push.128 u32lte assert
+    # => [num_inputs, inputs_ptr]
+
+    # get the number of word pairs to be hashed
+    u32divmod.8 swap
+    # => [num_inputs/8, num_inputs%8, inputs_ptr]
+
+    # get the end_addr for hash_memory_even proc
+    u32wrapping_mul.2 dup.2 u32wrapping_add movup.2
+    # => [inputs_ptr, end_addr, num_inputs%8]
+
+    # get the padding flag to add it to the capacity part
```

Comment:
I have only superficially reviewed the code below this line - but one thought: would it make sense to split this code into two parts:

1. Hashing of everything except for the last 2 words. For this, we could use `std::mem::pipe_double_words_to_memory`.
2. Hashing of the last 2 word.

The first part would be conditional - it would execute only if the number of inputs to hash is more than 8. The second part would be unconditional - it would always execute (and here we could also handle the case of 0 inputs).

I think the code could be much simpler if we do it this way (it should be possible to do the second part with a single while loop).

Reply:
I think generally I already do the same, but using `std::crypto::hashes::native::hash_memory_even` for hashing pairs of words.

I think I named the `end_addr` variable badly, since it's not really the `end_addr`, but the address where the integer number of pairs of words will end, so I can use this `end_addr` in the `hash_memory_even`. Also I think I know how to reduce the number of cycles to just one.